### PR TITLE
Use an explicit null check for null receivers, rather than an NPE

### PR DESCRIPTION
**Pull request by thecoop**
This is for #91236

### CODE REVIEW

File: modules/lang-painless/src/test/java/org/elasticsearch/painless/EqualsTests.java [hunk 1]
```diff
@@ -200,6 +202,18 @@ public void testStringEqualsMethodCall() {
...
+    }
+
+    public void testEqualsNullCheck() {
+        // get the same callsite working once, then with a null
+        // need to specify call site depth as 0 to force MIC to execute
+        assertEquals(false, exec("""
+            def list = ["foo", "foo", "quux", null];
+            def result = null;
+            for (int i=0; i<list.length; i+=2) {
+                result = list[i] == list[i+1];
```
[the remark this answers is not in the document]
Comment (thecoop) [reply modules/lang-painless/src/test/java/org/elasticsearch/painless/EqualsTests.java]: Done

**Comment by thecoop** [comment]
@elasticsearchmachine rerun elasticsearch-ci/bwc

**Review by jdconrad — COMMENTED** [review]
It's possible this has non-trivial performance implications. Do we have any benchmarks?

**Comment by stu-elastic** [comment]
Test lgtm, maybe we can do a rally run before/after?

**Comment by jdconrad** [comment]
Given this would be in the hot path for a large number of scripts, I definitely want to make sure we don't cause a large regression.

**Comment by thecoop** [comment]
I've done a rally run (http_logs, append-index-only-with-ingest-pipeline), this change looks to be a bit faster than baseline (5%)

**Review by stu-elastic — APPROVED** [review]
Great change, nice unit and perf test.

**Pull request merged**